Generator/Sources/NeedleFramework/Parsing/AbstractDependencyGraphParser.swift

```diff
@@ -47,17 +48,30 @@ class AbstractDependencyGraphParser {
     /// - throws: If any error occurred during execution.
     func executeAndCollectTaskHandles<ResultType>(with rootUrls: [URL], sourcesListFormatValue: String?, execution: (URL) -> SequenceExecutionHandle<ResultType>) throws -> [(SequenceExecutionHandle<ResultType>, URL)] {
         var urlHandles = [(SequenceExecutionHandle<ResultType>, URL)]()
-
+        
         // Enumerate all files and execute parsing sequences concurrently.
+        try collectFileUrlsIfNeeded(with: rootUrls, sourcesListFormatValue: sourcesListFormatValue)
+
+        for fileUrl in allFileUrls {
+            let taskHandle = execution(fileUrl)
+            urlHandles.append((taskHandle, fileUrl))
+        }
+
+        return urlHandles
+    }
+    
+    private func collectFileUrlsIfNeeded(with rootUrls: [URL], sourcesListFormatValue: String?) throws {
+        guard allFileUrls.isEmpty else {
+            return
+        }
+        
         let enumerator = FileEnumerator()
+        
         for url in rootUrls {
             try enumerator.enumerate(from: url, withSourcesListFormat: sourcesListFormatValue) { (fileUrl: URL) in
-                let taskHandle = execution(fileUrl)
-                urlHandles.append((taskHandle, fileUrl))
+                allFileUrls.append(fileUrl)
```

Comment:
You can also use the filepath string directly and avoid the url conversion cost altogether.  SwiftSyntax parser can be extended to take in a path string instead of a url. See https://github.com/uber/mockolo/blob/master/Sources/MockoloFramework/Parsers/ViaSwiftSyntax/SwiftSyntaxExtensions.swift#L23 for reference.

Reply:
True, rudro brought up the same point offline. But there's quite a few places in needle that depends on URL (CachedFileReader, URL filtering, etc..) that I decided it's not worth the cost of switching.

Reply:
That's fine; can revisit the conversion if it's still slow after optimizing other parts.